lib/replicache.dart

```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
 
-  /// Synchronizes the database with the server. New local transactions that have been executed since the last
-  /// sync are sent to the server, and new remote transactions are received and replayed.
+  Future<void> _sync() async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final syncHead = await _beginSync();
+    await _maybeEndSync(syncHead);
+  }
+
+  Future<String> _beginSync() async {
+    await _opened;
+    final beginSyncResult = await _invoke(_name, 'beginSync', {
+      'batchPushURL': _batchUrl,
+      // TODO(arv): Rename this param.
+      'clientViewURL': _remote,
+      'dataLayerAuth': _clientViewAuth,
+    });
+
+    final syncInfo = beginSyncResult['syncInfo'];
+
+    void checkStatus(Map<String, dynamic> data, String serverName) {
+      final httpStatusCode = data['httpStatusCode'];
+      switch (httpStatusCode) {
+        case HttpStatus.ok:
+          break;
+        case HttpStatus.unauthorized:
+          // TODO: Reauth
+          break;
+        default:
+          print(
+              'Got unexpected response from $serverName server: $httpStatusCode: ${data['errorMessage']}');
+      }
+    }
+
+    final batchPushInfo = syncInfo['batchPushInfo'];
+    if (batchPushInfo != null) {
+      checkStatus(batchPushInfo, 'batch');
+      final mutationInfos = batchPushInfo['batchPushResponse']['mutationInfos'];
+      if (mutationInfos != null) {
+        for (final mutationInfo in mutationInfos) {
+          print(
+              'MutationInfo: ID: ${mutationInfo['id']}, Error: ${mutationInfo['error']}');
+        }
+      }
+    }
+
+    checkStatus(syncInfo['clientViewInfo'], 'client view');
+
+    return beginSyncResult['syncHead'];
+  }
+
+  Future<void> _maybeEndSync(String syncHead) async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final res = await _invoke(_name, 'maybeEndSync', {'syncHead': syncHead});
+    final replayMutations = res['replayMutations'];
+    if (replayMutations == null || replayMutations.isEmpty) {
+      // All done.
+      await _checkChange(syncHead);
+      return;
+    }
+
+    // Replay.
+    for (final Map<String, dynamic> mutation in replayMutations) {
+      syncHead = await _replay(
+        syncHead,
+        mutation['original'],
+        mutation['name'],
+        mutation['args'],
+      );
+    }
+
+    await _maybeEndSync(syncHead);
+  }
+
+  Future<String> _replay<R, A>(
+    String basis,
+    String original,
+    String name,
+    A args,
+  ) async {
+    final mutatorImpl = _mutatorRegistry[name];
+    final res = await _mutate(name, mutatorImpl, args, invokeArgs: {
+      'rebaseOpts': {
+        'basis': basis,
+        'original': original,
+      }
+    });
+    return res.ref;
+  }
+
+  /// Synchronizes the database with the server. New local transactions that
+  /// have been executed since the last sync are sent to the server, and new
+  /// remote transactions are received and replayed.
```

Comment:
i would not say that new remote transactions are received and replayed. i would say that new state from the client view is received and any local mutations that are not included in that client view are replayed.

Reply:
This was my original phrasing, and I agree with your critique. Let me try again... these comments are worth fretting over since they will show up in e.g., autocomplete when people are using Replicache.

> Synchronizes this cache with the server. New local mutations are sent to the server, and the latest server state is applied to the cache. Any local mutations not included in the new server state are replayed. See the Replicache design document for more information on sync: https://github.com/rocicorp/replicache/blob/master/design.md